components/src/core/components/Input/TreeSelect/TreeSelect.vue

```diff
@@ -0,0 +1,545 @@
+              getUniqueIds(option.children!);
+            }
+          }
+        }
+        if (!(uniqueIds.size === idsCount)) {
+          console.warn(
+            'prop validation warning: options array prop cannot include non-unique ids',
+          );
+        }
+        return uniqueIds.size === idsCount;
+      },
+    },
+    preSelectedIds: {
+      type: Array as PropType<string[]>,
+      required: false,
+      default: [],
+    },
+    selectParentsOnChildSelection: {
+      type: Boolean,
+      default: false,
+    },
+    dropdownPosition: {
+      type: String,
+      default: BOTTOM,
+      validator: function(value: Position) {
+        return DROPDOWN_POSITIONS.indexOf(value) !== -1;
+      },
+    },
+  },
+
+  setup: function(props, {emit}) {
+    const expandedIds = ref<string[]>([]);
+    const selectedIdsObject = ref<SelectedIdsObject>({});
+    const optionsArr = ref<any[]>([...props.options]);
+    const dropdownOpen = ref<Boolean>(false);
+    const isAllSelected = ref<Boolean>(false);
+    const placeholderOption = ref<OptionProp | null>(null);
+
+    const levelizeOptions = (
+      optionsArr: Option[],
+      level: number,
+      parentOptions: Option[],
+    ) => {
+      for (let option of optionsArr) {
+        option['level'] = level;
+        option['parentOptions'] = parentOptions;
+        selectedIdsObject.value[option.id] = false; //initially makes all the ids false (unselected)
+        if (option.children ? option.children.length != 0 : false) {
+          levelizeOptions(option.children, level + 1, [
+            ...parentOptions,
+            option,
+          ]);
+        }
+      }
+    };
+    levelizeOptions(optionsArr.value, 1, []);
+    console.log(optionsArr.value, 'aftr lvlzed');
+    console.log(selectedIdsObject.value, 'selectedIdsObject.value');
+
+    const selectedIdsComputed = computed(() => {
+      let selectedIds = Object.keys(selectedIdsObject.value).filter(
+        k => selectedIdsObject.value[k],
+      );
+      return selectedIds;
+    });
+
+    const selectOptionsOnCheckbox = (
+      checkboxVal: Boolean,
+      optionId: string,
+    ) => {
+      let option = optionsArr.value.find((obj: {id: string}) => {
+        return obj.id == optionId;
+      });
+
+      if (checkboxVal) {
+        addChildrenToSelectedIdsArray(option);
+        if (props.selectParentsOnChildSelection) {
+          selectParentsIfAllChildrenAreSelected(option);
+        }
+        if (isIfAllOptionsSelected()) {
+          isAllSelected.value = true;
+        }
+      } else {
+        removeChildrenFromSelectedIdsArray(option);
+        if (props.selectParentsOnChildSelection) {
+          removeParentsOfUnselectedOptionsFromSelectedIdsArray(option);
+        }
+        isAllSelected.value = false;
+      }
+      emitSelectedIds();
+    };
+
+    const emitSelectedIds = () => {
+      emit('updateSelectedIds', selectedIdsComputed.value);
```

Comment:
Use a consistent name for the emitted event. What does other widgets uses there. is it update:selectedIds?